proposals/0000-th-stage-hygiene.rst

```diff
@@ -0,0 +1,369 @@
+   The top level is always stage 0.
+   A consequence of the above is all non-TH syntax in is also stage 0.
+
+#. Redefine quoting and splicing as acting on adjacent stages. Specifically, quoting quotes code from the next stage:
+   ::
+     G ⊢(n + 1) syntax
+     -----------------------
+     G ⊢(n) [| syntax |]
+   and splicing splices code from the previous stage:
+   ::
+     G ⊢(n - 1) syntax
+     -----------------------
+     G ⊢(n) $(syntax)
+
+   The existing side conditions, which restricting nested quotes and splices (i.e. stages outside of -1, 0, and 1) remain in place, but are ripe for removal in https://github.com/ghc-proposals/ghc-proposals/pulls/204.
+
+#. Add new syntax for stage-offset imports and bindings:
+   ::
+     <impdecl> ::= $import <integer-literal> <<existing syntax>>
+   This means import a module in stage *n* instead of stage 0 as per normal.
+   ::
+     <decl> ::= $let <integer-literal> <<existing syntax>> = <<existing syntax>>
+   In both case the ``$`` must not be followed by whitespace, both to avoid conflicts with other syntax and to be consistent with splices.
+   The means bind identifiers in stage *n* instead of stage 0 as per normal.
+   Module exports however are restricted to stage 0.
+
+#. The current "stage restriction" on splices using items from module is abolished.
+   Any stage n + 1 binding in a stage n splice is fair game.
+
+#. Relax ``-XTemplateHaskellQuotes`` to instead allow Template Haskell constructs, but restrict their usage so all syntax is in stages >= 0.
+
+#. Introduce ``-XTemplateStagePersistence``.
+   Which is implied by ``-XTemplateHaskellQuotes`` (and thus plain ``-XTemplateHaskell``) for backwards compatibility.
+   It allows the current behavior where we blur the distinction between stages.
+   In particular, with this enabled:
+
+   - Stage 0 identifiers bound in another module can be used in stage 1 (quotes) and stage -1 (splices).
+   - Stage 0 identifiers bound anywhere can be used in stage 1, and are automatically.
+   - Typed template haskell is allowed.
+   - The ``Lift`` type class and all its associated definitions are made available.
+
+   These are all already always permitted today.
+   But with ``-XNoTemplateStagePersistence``, overriding the default, all of those are *disabled*.
+
+#. Extend the command line [TODO bikeshed!!] with a way to specify per-stage package dependencies and the like.
+   If/when GHC becomes multi-target, by default stages >= 0 take GHC's target platform / the packages host platform (where compiled code runs), while stages < 0 take GHC's host platform / the packages build platform (where GHC runs).
+   But, the emitted platform can still be specified per-stage like the other flags.
+   This is needed when building TH functions to be used from cross compiled code.
+
+#. Add a Core "way" to GHC, which basically amounts to `-fexpose-all-unfoldings` and doing no work no past core [TODO bikeshed/clarify].
```

Comment:
This sounds like another use for https://gitlab.haskell.org/ghc/ghc/issues/10871.

Reply:
Awesome!! Commented on that issue, and will incorporate it into this.

Reply:
Updated the text. Hope it looks good now re this, and is more honest about what needs to be built as we discussed on IRC!